rust/chains/hyperlane-ethereum/src/rpc_clients/fallback.rs

```diff
@@ -50,16 +102,82 @@ impl<T> FallbackProvider<T> {
     }
 }
 
+impl<C> FallbackProvider<C>
+where
+    C: JsonRpcClient,
+{
+    async fn handle_stalled_provider(
+        &self,
+        priority: &PrioritizedProviderInner,
+        provider: &C,
+    ) -> Result<(), ProviderError> {
+        let now = Instant::now();
+        if now
+            .duration_since(priority.last_block_height.1)
+            .le(&self.max_block_time)
+        {
+            // Do nothing, it's too early to tell if the provider has stalled
+            return Ok(());
+        }
+
+        let current_block_height: u64 = provider
+            .request(BLOCK_NUMBER_RPC, ())
+            .await
+            .map(|r: U64| r.as_u64())
+            .unwrap_or(priority.last_block_height.0);
+        if current_block_height <= priority.last_block_height.0 {
+            // The `max_block_time` elapsed but the block number returned by the provider has not increased
+            self.deprioritize_provider(priority, provider).await;
+        } else {
+            self.update_last_seen_block(priority.index, current_block_height)
+                .await;
+        }
+        Ok(())
+    }
+
+    async fn deprioritize_provider(&self, priority: &PrioritizedProviderInner, provider: &C) {
```

Comment:
I moved the log to the call site of `deprioritize_provider` which is fine since it's only one site, otherwise I would've probably kept the `provider` argument to reduce logging duplication